lib/raygun.messageBuilder.ts

```diff
@@ -0,0 +1,248 @@
+/*
+ * raygun
+ * https://github.com/MindscapeHQ/raygun4node
+ *
+ * Copyright (c) 2015 MindscapeHQ
+ * Licensed under the MIT license.
+ */
+
+"use strict";
+
+import os from "os";
+import stackTrace from "stack-trace";
+
+import {
+  MessageBuilderOptions,
+  StackFrame,
+  Message,
+  MessageBuilding,
+  IndexableError,
+  RawUserData,
+  UserDetails,
+  RequestParams,
+  Tag,
+  CustomData,
+  Environment,
+  BuiltError,
+} from "./types";
+
+type UserMessageData = RawUserData | string | undefined;
+
+const humanString = require("object-to-human-string");
+const packageDetails = require("../package.json");
+
+function filterKeys(obj: { [key: string]: any }, filters: string[]) {
+  if (!obj || !filters || typeof obj !== "object") {
+    return obj;
+  }
+  Object.keys(obj).forEach(function (i) {
+    if (filters.indexOf(i) > -1) {
+      delete obj[i];
+    } else {
+      obj[i] = filterKeys(obj[i], filters);
+    }
+  });
+  return obj;
+}
+
+function getStackTrace(
+  error: Error,
+  options: MessageBuilderOptions
+): StackFrame[] {
+  const stack: StackFrame[] = [];
+  const trace = stackTrace.parse(error);
+
+  trace.forEach(function (callSite) {
+    const frame: StackFrame = {
+      lineNumber: callSite.getLineNumber(),
+      className: callSite.getTypeName() || "unknown",
+      fileName: callSite.getFileName(),
+      methodName: callSite.getFunctionName() || "[anonymous]",
+    };
+
+    if (
+      !!options.reportColumnNumbers &&
+      typeof callSite.getColumnNumber === "function"
+    ) {
+      frame.columnNumber = callSite.getColumnNumber();
+    }
+
+    stack.push(frame);
+  });
+
+  return stack;
+}
+
+function buildError(
+  error: IndexableError,
+  options: MessageBuilderOptions
+): BuiltError {
+  const builtError: BuiltError = {
+    stackTrace: getStackTrace(error, options),
+    message: error.message || "NoMessage",
+    className: error.name,
+  };
+
+  let innerError: Error | undefined = undefined;
+
+  if (options.innerErrorFieldName) {
+    innerError =
+      typeof error[options.innerErrorFieldName!] === "function"
+        ? error[options.innerErrorFieldName!]()
+        : error[options.innerErrorFieldName!];
+  }
+
+  if (innerError instanceof Error) {
+    builtError.innerError = buildError(innerError, options);
+  }
+
+  return builtError;
+}
+
+export class RaygunMessageBuilder {
+  _filters: string[];
+  options: MessageBuilderOptions;
+  message: MessageBuilding;
+
+  constructor(options: MessageBuilderOptions = {}) {
+    options = options || {};
+    this.options = options;
+    this._filters = options.filters || [];
```

Comment:
Since this is an internal class we can rely on the type checker to ensure that `options.filters` is an array, so we don't need to handle those cases at runtime anymore, so it should be okay :)